handler/metadata_test.go

```diff
@@ -29,19 +32,29 @@ func TestMetadata(t *testing.T) {
 		ts := httptest.NewServer(r)
 		defer ts.Close()
 
-		Convey("when a request is made to /metadata", func() {
-			go func(){
-				b := <-producer.Channels().Output
-				t.Log(string(b))
-			}()
-			resp := testDoRequest(t, ts, http.MethodPost, "/metadata", bytes.NewReader([]byte("{\"\":\"\"}")))
+		Convey("when a request is made to /metadata", func(c C) {
+			go func(c C){
+				body := []byte(`{"dataset_id":"cantabular-example-1","edition":"2021","version":1}`)
+				resp := testDoRequest(t, ts, http.MethodPost, "/metadata", bytes.NewReader(body))
+
+				c.Convey("The the returned status code should equal 202 Status Accepted", func(c C) {
```

Comment:
When running unit tests locally, a race condition was identified at this point.
This can be fixed by adding a wait group and waiting for it just after reading from the output channel, and only checking values afterwards. Tested this locally:

```
		Convey("when a request is made to /metadata", func(c C) {
			wg := &sync.WaitGroup{}
			wg.Add(1)
			go func(c C) {
				defer wg.Done()
				body := []byte(`{"dataset_id":"cantabular-example-1","edition":"2021","version":1}`)
				resp := testDoRequest(t, ts, http.MethodPost, "/metadata", bytes.NewReader(body))

				c.Convey("The the returned status code should equal 202 Status Accepted", func(c C) {
					c.So(resp.StatusCode, ShouldEqual, http.StatusAccepted)
				})
			}(c)

			expected := event.CantabularMetadataExport{
				DatasetID: "cantabular-example-1",
				Edition:   "2021",
				Version:   1,
			}
			Convey("And the expected message is produced", func() {
				b := <-producer.Channels().Output
				wg.Wait()
				var got event.CantabularMetadataExport
				err := schema.CantabularMetadataExport.Unmarshal(b, &got)
				So(err, ShouldBeNil)
				So(got, ShouldResemble, expected)
			})
		})
```

Reply:
Ok, have added the waitgroup to the test. In regards to the other comments, the main thing I don't like about having the handlers in another package is that everything has to be called foobarHandler and it gets unwieldy. I guess it's not such a big deal in a small service like this but the bigger ones get really messy. The buildRoutes function can't be in the api package either as it's on the Service struct (to prevent having to pass 50 arguments to it) and would cause a dependency cycle.